### DISCUSSION TITLE
lock bit using swwe - how to implement

### DISCUSSION THREAD

**Discussion by arnonsha** [SystemRDL/systemrdl-compiler#189] [Q&A] [answered]
I am trying to implement a locking bit called LOCK, which is part of my register file. when LOCK = 1 , writing to all other registers is disabled.
I am trying to do that using the swwe attribute. My rdl file looks like this:

![image](https://github.com/SystemRDL/systemrdl-compiler/assets/35176391/615ea66d-720a-4de7-be73-6ba12a2cc3de)

When running peakrdl I get the following error:

```ssue.rdl:24:20: fatal: Reference to 'LOCKREG' not found
            swwe = LOCKREG.LOCK;
```

I get this error also when I try swwe=LOCK

What am I doing wrong ?

**Comment by amykyta3** [comment, marked as answer]
Due to how namespace scopes work in SystemRDL, you cannot access `LOCKREG` directly from within the lexical scope of `CSR1` or `csr`. Instead, try using a dynamic property assignment outside of the blocks:

```systemrdl
reg {
    field {} LOCK[0:0] = 0;
} LOCKREG;

reg {
    field {} csr = 0;
} CSR1;

CSR1.csr->swwe = LOCKREG.LOCK;
```

Also unrelated - I see you are creating an alternate reset signal, presumably to override the default reset to be asynchronous. PeakRDL-regblock lets you set the default reset via the command-line (`peakrdl regblock --default-reset arst ...`) or via the tool's [TOML config](https://peakrdl-regblock.readthedocs.io/en/latest/configuring.html).

**Reply by arnonsha** [reply]
thank you @amykyta3 

Indeed this is working, however I have 10s of fields that I need to lock. Sounds like I'll need a dedicated line for every field. I tried applying the lock bit on a whole register (CSR1->swwe = LOCKREG.LOCK) and got the following error:

`fatal: The property 'swwe' is not valid for 'reg' components`


Is there a more compact way to apply the locking on all the fields of a register ?

thanks

**Reply by amykyta3** [reply]
Yes you can apply the default assignment as follows:

```
reg {
    field {} LOCK[0:0] = 0;
} LOCKREG;

// Every field declared below here will be assigned swwe
default swwe = LOCKREG.LOCK;

reg {
    field {} csr_a = 0;
    field {} csr_b = 0;
    field {} csr_c = 0;
} CSR1;
```

**Reply by arnonsha** [reply]
Managed to make it work ! thank you for your support @amykyta3
